File: examples/src/test/scala/docs/eventsourced/ShoppingCart.scala
/*
 * Copyright (C) 2020-2025 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.eventsourced

import java.time.Instant

import scala.concurrent.duration._

import akka.actor.typed.ActorRef
import akka.actor.typed.Behavior
import akka.actor.typed.SupervisorStrategy
import akka.persistence.typed.PersistenceId
import akka.persistence.typed.scaladsl.Effect
import akka.persistence.typed.scaladsl.EventSourcedBehavior
import akka.persistence.typed.scaladsl.ReplyEffect
import akka.persistence.typed.scaladsl.RetentionCriteria
import akka.serialization.jackson.CborSerializable

//#imports
import akka.actor.typed.ActorSystem
import akka.cluster.sharding.typed.scaladsl.ClusterSharding
import akka.cluster.sharding.typed.scaladsl.Entity
import akka.cluster.sharding.typed.scaladsl.EntityTypeKey

//#imports

/**
 * This is an event sourced actor. It has a state, [[ShoppingCart.State]], which
 * stores the current shopping cart items and whether it's checked out.
 *
 * Event sourced actors are interacted with by sending them commands,
 * see classes implementing [[ShoppingCart.Command]].
 *
 * Commands get translated to events, see classes implementing [[ShoppingCart.Event]].
 * It's the events that get persisted by the entity. Each event will have an event handler
 * registered for it, and an event handler updates the current state based on the event.
 * This will be done when the event is first created, and it will also be done when the entity is
 * loaded from the database - each event will be replayed to recreate the state
 * of the entity.
 */
object ShoppingCart {

  /**
   * The current state held by the persistent entity.
   */
  final case class State(items: Map[String, Int], checkoutDate: Option[Instant]) extends CborSerializable {

    def isCheckedOut: Boolean =
      checkoutDate.isDefined

    def hasItem(itemId: String): Boolean =
      items.contains(itemId)

    def isEmpty: Boolean =
      items.isEmpty

    def updateItem(itemId: String, quantity: Int): State = {
      quantity match {
        case 0 => copy(items = items - itemId)
        case _ => copy(items = items + (itemId -> quantity))
      }
    }

    def removeItem(itemId: String): State =
      copy(items = items - itemId)

    def checkout(now: Instant): State =
      copy(checkoutDate = Some(now))

    def toSummary: Summary =
      Summary(items, isCheckedOut)
  }
  object State {
    val empty = State(items = Map.empty, checkoutDate = None)
  }

  /**
   * This interface defines all the commands that the ShoppingCart persistent actor supports.
   */
  sealed trait Command extends CborSerializable

  /**
   * A command to add an item to the cart.
   *
   * It can reply with `Confirmation`, which is sent back to the caller when
   * all the events emitted by this command are successfully persisted.
   */
  final case class AddItem(itemId: String, quantity: Int, replyTo: ActorRef[Confirmation]) extends Command

  /**
   * A command to remove an item from the cart.
   */
  final case class RemoveItem(itemId: String, replyTo: ActorRef[Confirmation]) extends Command

  /**
   * A command to adjust the quantity of an item in the cart.
   */
  final case class AdjustItemQuantity(itemId: String, quantity: Int, replyTo: ActorRef[Confirmation]) extends Command

  /**
   * A command to checkout the shopping cart.
   */
  final case class Checkout(replyTo: ActorRef[Confirmation]) extends Command

  /**
   * A command to get the current state of the shopping cart.
   */
  final case class Get(replyTo: ActorRef[Summary]) extends Command

  /**
   * Summary of the shopping cart state, used in reply messages.
   */
  final case class Summary(items: Map[String, Int], checkedOut: Boolean) extends CborSerializable

  sealed trait Confirmation extends CborSerializable

  final case class Accepted(summary: Summary) extends Confirmation

  final case class Rejected(reason: String) extends Confirmation

  /**
   * This interface defines all the events that the ShoppingCart supports.
   */
  sealed trait Event extends CborSerializable {
    def cartId: String
  }

  final case class ItemAdded(cartId: String, itemId: String, quantity: Int) extends Event

  final case class ItemRemoved(cartId: String, itemId: String) extends Event

  final case class ItemQuantityAdjusted(cartId: String, itemId: String, newQuantity: Int) extends Event

  final case class CheckedOut(cartId: String, eventTime: Instant) extends Event

  //#slicingTags
  val tags = Vector.tabulate(5)(i => s"carts-$i")
  //#slicingTags

  //#tagging
  val EntityKey: EntityTypeKey[Command] = EntityTypeKey[Command]("ShoppingCart")

  def init(system: ActorSystem[_]): Unit = {
    ClusterSharding(system).init(Entity(EntityKey) { entityContext =>
      val i = math.abs(entityContext.entityId.hashCode % tags.size)
      val selectedTag = tags(i)
      ShoppingCart(entityContext.entityId, selectedTag)
    }.withRole("write-model"))
  }

  def apply(cartId: String, projectionTag: String): Behavior[Command] = {
    EventSourcedBehavior
      .withEnforcedReplies[Command, Event, State](
        PersistenceId(EntityKey.name, cartId),
        State.empty,
        (state, command) =>
          //The shopping cart behavior changes if it's checked out or not.
          // The commands are handled differently for each case.
          if (state.isCheckedOut) checkedOutShoppingCart(cartId, state, command)
          else openShoppingCart(cartId, state, command),
        (state, event) => handleEvent(state, event))
      .withTagger(_ => Set(projectionTag))
      .withRetention(RetentionCriteria.snapshotEvery(numberOfEvents = 100))
      .onPersistFailure(SupervisorStrategy.restartWithBackoff(200.millis, 5.seconds, 0.1))
  }
  //#tagging

  private def openShoppingCart(cartId: String, state: State, command: Command): ReplyEffect[Event, State] =
    command match {
      case AddItem(itemId, quantity, replyTo) =>
        if (state.hasItem(itemId))
          Effect.reply(replyTo)(Rejected(s"Item '$itemId' was already added to this shopping cart"))
        else if (quantity <= 0)
          Effect.reply(replyTo)(Rejected("Quantity must be greater than zero"))
        else
          Effect
            .persist(ItemAdded(cartId, itemId, quantity))
            .thenReply(replyTo)(updatedCart => Accepted(updatedCart.toSummary))

      case RemoveItem(itemId, replyTo) =>
        if (state.hasItem(itemId))
          Effect.persist(ItemRemoved(cartId, itemId)).thenReply(replyTo)(updatedCart => Accepted(updatedCart.toSummary))
        else
          Effect.reply(replyTo)(Accepted(state.toSummary)) // removing an item is idempotent

      case AdjustItemQuantity(itemId, quantity, replyTo) =>
        if (quantity <= 0)
          Effect.reply(replyTo)(Rejected("Quantity must be greater than zero"))
        else if (state.hasItem(itemId))
          Effect
            .persist(ItemQuantityAdjusted(cartId, itemId, quantity))
            .thenReply(replyTo)(updatedCart => Accepted(updatedCart.toSummary))
        else
          Effect.reply(replyTo)(Rejected(s"Cannot adjust quantity for item '$itemId'. Item not present on cart"))

      case Checkout(replyTo) =>
        if (state.isEmpty)
          Effect.reply(replyTo)(Rejected("Cannot checkout an empty shopping cart"))
        else
          Effect
            .persist(CheckedOut(cartId, Instant.now()))
            .thenReply(replyTo)(updatedCart => Accepted(updatedCart.toSummary))

      case Get(replyTo) =>
        Effect.reply(replyTo)(state.toSummary)
    }

  private def checkedOutShoppingCart(cartId: String, state: State, command: Command): ReplyEffect[Event, State] =
    command match {
      case Get(replyTo) =>
        Effect.reply(replyTo)(state.toSummary)
      case cmd: AddItem =>
        Effect.reply(cmd.replyTo)(
          Rejected(s"Can't add an item to an already checked out shopping cart with id [$cartId]"))
      case cmd: RemoveItem =>
        Effect.reply(cmd.replyTo)(
          Rejected(s"Can't remove an item from an already checked out shopping cart with id [$cartId]"))
      case cmd: AdjustItemQuantity =>
        Effect.reply(cmd.replyTo)(
          Rejected(s"Can't adjust item on an already checked out shopping cart with id [$cartId]"))
      case cmd: Checkout =>
        Effect.reply(cmd.replyTo)(Rejected(s"Can't checkout already checked out shopping cart with id [$cartId]"))
    }

  private def handleEvent(state: State, event: Event) = {
    event match {
      case ItemAdded(_, itemId, quantity)            => state.updateItem(itemId, quantity)
      case ItemRemoved(_, itemId)                    => state.removeItem(itemId)
      case ItemQuantityAdjusted(_, itemId, quantity) => state.updateItem(itemId, quantity)
      case CheckedOut(_, eventTime)                  => state.checkout(eventTime)
    }
  }
}

File: examples/src/test/java/jdocs/eventsourced/ShoppingCart.java
/*
 * Copyright (C) 2020-2025 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.eventsourced;

import akka.actor.typed.ActorRef;
import akka.actor.typed.ActorSystem;
import akka.actor.typed.Behavior;
import akka.actor.typed.SupervisorStrategy;
import akka.cluster.sharding.typed.javadsl.ClusterSharding;
import akka.cluster.sharding.typed.javadsl.Entity;
import akka.cluster.sharding.typed.javadsl.EntityTypeKey;
import akka.persistence.typed.PersistenceId;
import akka.persistence.typed.javadsl.CommandHandlerWithReply;
import akka.persistence.typed.javadsl.CommandHandlerWithReplyBuilder;
import akka.persistence.typed.javadsl.EventHandler;
import akka.persistence.typed.javadsl.EventSourcedBehaviorWithEnforcedReplies;
import akka.persistence.typed.javadsl.ReplyEffect;
import akka.persistence.typed.javadsl.RetentionCriteria;
import akka.serialization.jackson.CborSerializable;
import com.fasterxml.jackson.annotation.JsonCreator;

import java.time.Duration;
import java.time.Instant;
import java.util.*;

/**
 * This is an event sourced actor. It has a state, {@link ShoppingCart.State}, which stores the
 * current shopping cart items and whether it's checked out.
 *
 * <p>Event sourced actors are interacted with by sending them commands, see classes implementing
 * {@link ShoppingCart.Command}.
 *
 * <p>Commands get translated to events, see classes implementing {@link ShoppingCart.Event}. It's
 * the events that get persisted by the entity. Each event will have an event handler registered for
 * it, and an event handler updates the current state based on the event. This will be done when the
 * event is first created, and it will also be done when the entity is loaded from the database -
 * each event will be replayed to recreate the state of the entity.
 */
public class ShoppingCart
    extends EventSourcedBehaviorWithEnforcedReplies<
        ShoppingCart.Command, ShoppingCart.Event, ShoppingCart.State> {

  /** The state for the {@link ShoppingCart} entity. */
  public final class State implements CborSerializable {
    private Map<String, Integer> items = new HashMap<>();
    private Optional<Instant> checkoutDate = Optional.empty();

    public boolean isCheckedOut() {
      return checkoutDate.isPresent();
    }

    public Optional<Instant> getCheckoutDate() {
      return checkoutDate;
    }

    public boolean isEmpty() {
      return items.isEmpty();
    }

    public boolean hasItem(String itemId) {
      return items.containsKey(itemId);
    }

    public State updateItem(String itemId, int quantity) {
      if (quantity == 0) {
        items.remove(itemId);
      } else {
        items.put(itemId, quantity);
      }
      return this;
    }

    public State removeItem(String itemId) {
      items.remove(itemId);
      return this;
    }

    public State checkout(Instant now) {
      checkoutDate = Optional.of(now);
      return this;
    }

    public Summary toSummary() {
      return new Summary(items, isCheckedOut());
    }
  }

  /** This interface defines all the commands that the ShoppingCart persistent actor supports. */
  public interface Command extends CborSerializable {}

  /**
   * A command to add an item to the cart.
   *
   * <p>It can reply with `Confirmation`, which is sent back to the caller when all the events
   * emitted by this command are successfully persisted.
   */
  public static class AddItem implements Command {
    public final String itemId;
    public final int quantity;
    public final ActorRef<Confirmation> replyTo;

    public AddItem(String itemId, int quantity, ActorRef<Confirmation> replyTo) {
      this.itemId = itemId;
      this.quantity = quantity;
      this.replyTo = replyTo;
    }
  }

  /** A command to remove an item from the cart. */
  public static class RemoveItem implements Command {
    public final String itemId;
    public final ActorRef<Confirmation> replyTo;

    @JsonCreator
    public RemoveItem(String itemId, ActorRef<Confirmation> replyTo) {
      this.itemId = itemId;
      this.replyTo = replyTo;
    }
  }

  /** A command to adjust the quantity of an item in the cart. */
  public static class AdjustItemQuantity implements Command {
    public final String itemId;
    public final int quantity;
    public final ActorRef<Confirmation> replyTo;

    public AdjustItemQuantity(String itemId, int quantity, ActorRef<Confirmation> replyTo) {
      this.itemId = itemId;
      this.quantity = quantity;
      this.replyTo = replyTo;
    }
  }

  /**
   * A command to get the current state of the shopping cart.
   *
   * <p>The reply type is the {@link Summary}
   */
  public static class Get implements Command {
    public final ActorRef<Summary> replyTo;

    @JsonCreator
    public Get(ActorRef<Summary> replyTo) {
      this.replyTo = replyTo;
    }
  }

  /**
   * A command to checkout the shopping cart.
   *
   * <p>The reply type is the {@link Confirmation}, which will be returned when the events have been
   * emitted.
   */
  public static class Checkout implements Command {
    public final ActorRef<Confirmation> replyTo;

    @JsonCreator
    public Checkout(ActorRef<Confirmation> replyTo) {
      this.replyTo = replyTo;
    }
  }

  /** Summary of the shopping cart state, used in reply messages. */
  public static final class Summary implements CborSerializable {
    public final Map<String, Integer> items;
    public final boolean checkedOut;

    public Summary(Map<String, Integer> items, boolean checkedOut) {
      // Summary is included in messages and should therefore be immutable
      this.items = Collections.unmodifiableMap(new HashMap<>(items));
      this.checkedOut = checkedOut;
    }
  }

  public interface Confirmation extends CborSerializable {}

  public static class Accepted implements Confirmation {
    public final Summary summary;

    @JsonCreator
    public Accepted(Summary summary) {
      this.summary = summary;
    }
  }

  public static class Rejected implements Confirmation {
    public final String reason;

    public Rejected(String reason) {
      this.reason = reason;
    }
  }

  public interface Event extends CborSerializable {
    public String getCartId();
  }

  public static final class ItemAdded implements Event {
    public final String cartId;
    public final String itemId;
    public final int quantity;

    public ItemAdded(String cartId, String itemId, int quantity) {
      this.cartId = cartId;
      this.itemId = itemId;
      this.quantity = quantity;
    }

    public String getCartId() {
      return cartId;
    }

    @Override
    public String toString() {
      return "ItemAdded(" + cartId + "," + itemId + "," + quantity + ")";
    }
  }

  public static final class ItemRemoved implements Event {
    public final String cartId;
    public final String itemId;

    public ItemRemoved(String cartId, String itemId) {
      this.cartId = cartId;
      this.itemId = itemId;
    }

    public String getCartId() {
      return cartId;
    }

    @Override
    public String toString() {
      return "ItemRemoved(" + cartId + "," + itemId + ")";
    }
  }

  public static final class ItemQuantityAdjusted implements Event {
    public final String cartId;
    public final String itemId;
    public final int quantity;

    public ItemQuantityAdjusted(String cartId, String itemId, int quantity) {
      this.cartId = cartId;
      this.itemId = itemId;
      this.quantity = quantity;
    }

    public String getCartId() {
      return cartId;
    }

    @Override
    public String toString() {
      return "ItemQuantityAdjusted(" + cartId + "," + itemId + "," + quantity + ")";
    }
  }

  public static class CheckedOut implements Event {

    public final String cartId;
    public final Instant eventTime;

    public CheckedOut(String cartId, Instant eventTime) {
      this.cartId = cartId;
      this.eventTime = eventTime;
    }

    public String getCartId() {
      return cartId;
    }

    @Override
    public String toString() {
      return "CheckedOut(" + cartId + "," + eventTime + ")";
    }
  }

  public static EntityTypeKey<Command> ENTITY_TYPE_KEY =
      EntityTypeKey.create(Command.class, "ShoppingCart");

  public static void init(ActorSystem<?> system, int numberOfEventProcessors) {
    ClusterSharding.get(system)
        .init(
            Entity.of(
                    ENTITY_TYPE_KEY,
                    entityContext -> {
                      return ShoppingCart.create(entityContext.getEntityId());
                    })
                .withRole("write-model"));
  }

  public static Behavior<Command> create(String cartId) {
    return new ShoppingCart(cartId);
  }

  private final String cartId;

  private ShoppingCart(String cartId) {
    super(
        PersistenceId.of(ENTITY_TYPE_KEY.name(), cartId),
        SupervisorStrategy.restartWithBackoff(Duration.ofMillis(200), Duration.ofSeconds(5), 0.1));
    this.cartId = cartId;
  }

  // #slicingTags
  public static final List<String> tags =
      Collections.unmodifiableList(
          Arrays.asList("carts-0", "carts-1", "carts-2", "carts-3", "carts-4"));
  // #slicingTags

  // #tagging
  @Override
  public Set<String> tagsFor(Event event) {
    int n = Math.abs(cartId.hashCode() % tags.size());
    String selectedTag = tags.get(n);
    return Collections.singleton(selectedTag);
  }
  // #tagging

  @Override
  public State emptyState() {
    return new State();
  }

  private final CheckedOutCommandHandlers checkedOutCommandHandlers =
      new CheckedOutCommandHandlers();
  private final OpenShoppingCartCommandHandlers openShoppingCartCommandHandlers =
      new OpenShoppingCartCommandHandlers();

  @Override
  public CommandHandlerWithReply<Command, Event, State> commandHandler() {
    CommandHandlerWithReplyBuilder<Command, Event, State> b = newCommandHandlerWithReplyBuilder();

    b.forState(state -> !state.isCheckedOut())
        .onCommand(AddItem.class, openShoppingCartCommandHandlers::onAddItem)
        .onCommand(RemoveItem.class, openShoppingCartCommandHandlers::onRemoveItem)
        .onCommand(AdjustItemQuantity.class, openShoppingCartCommandHandlers::onAdjustItemQuantity)
        .onCommand(Checkout.class, openShoppingCartCommandHandlers::onCheckout);

    b.forState(state -> state.isCheckedOut())
        .onCommand(AddItem.class, checkedOutCommandHandlers::onAddItem)
        .onCommand(RemoveItem.class, checkedOutCommandHandlers::onRemoveItem)
        .onCommand(AdjustItemQuantity.class, checkedOutCommandHandlers::onAdjustItemQuantity)
        .onCommand(Checkout.class, checkedOutCommandHandlers::onCheckout);

    b.forAnyState().onCommand(Get.class, this::onGet);

    return b.build();
  }

  private ReplyEffect<Event, State> onGet(State state, Get cmd) {
    return Effect().reply(cmd.replyTo, state.toSummary());
  }

  private class OpenShoppingCartCommandHandlers {

    public ReplyEffect<Event, State> onAddItem(State state, AddItem cmd) {
      if (state.hasItem(cmd.itemId)) {
        return Effect()
            .reply(
                cmd.replyTo,
                new Rejected("Item '" + cmd.itemId + "' was already added to this shopping cart"));
      } else if (cmd.quantity <= 0) {
        return Effect().reply(cmd.replyTo, new Rejected("Quantity must be greater than zero"));
      } else {
        return Effect()
            .persist(new ItemAdded(cartId, cmd.itemId, cmd.quantity))
            .thenReply(cmd.replyTo, updatedCart -> new Accepted(updatedCart.toSummary()));
      }
    }

    public ReplyEffect<Event, State> onRemoveItem(State state, RemoveItem cmd) {
      if (state.hasItem(cmd.itemId)) {
        return Effect()
            .persist(new ItemRemoved(cartId, cmd.itemId))
            .thenReply(cmd.replyTo, updatedCart -> new Accepted(updatedCart.toSummary()));
      } else {
        return Effect().reply(cmd.replyTo, new Accepted(state.toSummary()));
      }
    }

    public ReplyEffect<Event, State> onAdjustItemQuantity(State state, AdjustItemQuantity cmd) {
      if (cmd.quantity <= 0) {
        return Effect().reply(cmd.replyTo, new Rejected("Quantity must be greater than zero"));
      } else if (state.hasItem(cmd.itemId)) {
        return Effect()
            .persist(new ItemQuantityAdjusted(cartId, cmd.itemId, cmd.quantity))
            .thenReply(cmd.replyTo, updatedCart -> new Accepted(updatedCart.toSummary()));
      } else {
        return Effect()
            .reply(
                cmd.replyTo,
                new Rejected(
                    "Cannot adjust quantity for item '"
                        + cmd.itemId
                        + "'. Item not present on cart"));
      }
    }

    public ReplyEffect<Event, State> onCheckout(State state, Checkout cmd) {
      if (state.isEmpty()) {
        return Effect().reply(cmd.replyTo, new Rejected("Cannot checkout an empty shopping cart"));
      } else {
        return Effect()
            .persist(new CheckedOut(cartId, Instant.now()))
            .thenReply(cmd.replyTo, updatedCart -> new Accepted(updatedCart.toSummary()));
      }
    }
  }

  private class CheckedOutCommandHandlers {
    ReplyEffect<Event, State> onAddItem(AddItem cmd) {
      return Effect()
          .reply(
              cmd.replyTo,
              new Rejected("Can't add an item to an already checked out shopping cart"));
    }

    ReplyEffect<Event, State> onRemoveItem(RemoveItem cmd) {
      return Effect()
          .reply(
              cmd.replyTo,
              new Rejected("Can't remove an item from an already checked out shopping cart"));
    }

    ReplyEffect<Event, State> onAdjustItemQuantity(AdjustItemQuantity cmd) {
      return Effect()
          .reply(
              cmd.replyTo,
              new Rejected("Can't adjust item on an already checked out shopping cart"));
    }

    ReplyEffect<Event, State> onCheckout(Checkout cmd) {
      return Effect()
          .reply(cmd.replyTo, new Rejected("Can't checkout already checked out shopping cart"));
    }
  }

  @Override
  public EventHandler<State, Event> eventHandler() {
    return newEventHandlerBuilder()
        .forAnyState()
        .onEvent(ItemAdded.class, (state, event) -> state.updateItem(event.itemId, event.quantity))
        .onEvent(ItemRemoved.class, (state, event) -> state.removeItem(event.itemId))
        .onEvent(
            ItemQuantityAdjusted.class,
            (state, event) -> state.updateItem(event.itemId, event.quantity))
        .onEvent(CheckedOut.class, (state, event) -> state.checkout(event.eventTime))
        .build();
  }

  @Override
  public RetentionCriteria retentionCriteria() {
    // enable snapshotting
    return RetentionCriteria.snapshotEvery(100, 3);
  }
}
